warehouse/ingest-core/src/main/java/datawave/ingest/table/aggregator/GlobalIndexUidAggregator.java

```diff
@@ -150,69 +155,113 @@ public Value reduce(Key key, Iterator<Value> iter) {
+                    log.debug("SeenIgnore is true. Skipping collections");
                 }
                 
-                // if delta > 0, we are collecting the uid list
-                // in the protobuf into our object's uid list.
-                if (delta > 0) {
-                    
+                // Partial major compactions make it possible for any combination of delta (positive/negative)
+                // and UIDs in the UID and REMOVEDUID list. For example, consider the case where uid1 and uid2
+                // are both added at separate times, and later uid1 is removed. If all three of those operations
+                // end up in different rfiles, then a partial major compaction could select the file with the add
+                // of uid2 and the removal of uid1 for compacting. The resulting protocol buffer would have a
+                // +1 for uid2 and a -1 for uid1, yielding a count of 0 and uid2 in the UID list and uid1 in the
+                // REMOVEDUID list. If another uid, uid3, ended up with the same order of operations as uid1
+                // then we could have a count of -1 (+1 for uid2, -1 for uid1, -1 for uid3) but still have uid2
+                // in the UID list and uid1 and uid3 in the REMOVEDUID list.
+                
+                // Manage quarantined and released UIDs.
+                if (delta > 0 && v.getQUARANTINEUIDCount() > 0 && v.getUIDCount() == 0 && v.getREMOVEDUIDCount() == 0 && v.getRELEASEUIDCount() == 0) {
+                    // This is for backwards compatibility. Previously the QUARANTINEDUID list was used for both
+                    // quarantining and releasing UIDs (negative count was quarantine and positive count was release).
+                    // However, this method does not work when a protocol buffer gets combined via a partial major
+                    // compaction. If we have only quarantined UIDs in the protocol buffer with a positive count, and
+                    // nothing else, then handle that as a released UID(s).
                     for (String uid : v.getQUARANTINEUIDList()) {
-                        
-                        quarantinedIds.remove(uid);
-                        releasedUids.add(uid);
+                        if ((timestampsIgnored || !quarantinedIds.contains(uid)) && !uids.contains(uid))
+                            releasedUids.add(uid);
+                        if (timestampsIgnored)
+                            quarantinedIds.remove(uid);
                     }
-                    
+                } else {
+                    // Add any quarantined UIDs to the internal list. If we're paying attention to timestamps, then we
+                    // don't want to quarantine any uid that we've already seen (i.e., with a newer timestamp) that has
+                    // been released.
+                    for (String uid : v.getQUARANTINEUIDList()) {
```

Comment:
I'm unaware of any uses of the quarantine and release capabilities.  Perhaps it would be best to deprecate and drop support, for the sake of simplifying this complicated and critical aggregator.  Thoughts?

Reply:
It would make the logic simpler, for sure.

Reply:
I will try removing that, and covering the duplicate UID cases.